datafusion/src/physical_plan/metrics.rs

```diff
@@ -0,0 +1,461 @@
+                )
+            })
+            .chain(self.labels().iter().cloned())
+            .peekable();
+
+        // print out the labels specially
+        if iter.peek().is_some() {
+            write!(f, "{{")?;
+
+            let mut is_first = true;
+            for i in iter {
+                if !is_first {
+                    write!(f, ", ")?;
+                } else {
+                    is_first = false;
+                }
+
+                write!(f, "{}", i)?;
+            }
+
+            write!(f, "}}")?;
+        }
+
+        // and now the value
+        let format_as_duration = match &self.kind {
+            MetricKind::OutputRows => false,
+            MetricKind::CPUTime => true,
+            MetricKind::Custom(name) => name.contains("Time") || name.contains("time"),
+        };
+
+        if format_as_duration {
+            let duration = std::time::Duration::from_nanos(self.value() as u64);
+            write!(f, "={:?}", duration)
+        } else {
+            write!(f, "={}", self.value())
+        }
+    }
+}
+
+impl SQLMetric {
+    /// Create a new SQLMetric
+    pub fn new(kind: MetricKind, partition: Option<usize>) -> Self {
+        Self {
+            value: 0.into(),
+            labels: vec![],
+            partition,
+            kind,
+        }
+    }
+
+    /// Add a new label to this metric
+    pub fn new_with_labels(
+        kind: MetricKind,
+        partition: Option<usize>,
+        labels: Vec<Label>,
+    ) -> Self {
+        Self {
+            value: 0.into(),
+            labels,
+            partition,
+            kind,
+        }
+    }
+
+    /// Add a new label to this metric
+    pub fn with(mut self, label: Label) -> Self {
+        self.labels.push(label);
+        self
+    }
+
+    /// Add the standard name for output rows
+
+    /// Get the current value
+    pub fn value(&self) -> usize {
+        self.value.load(Ordering::Relaxed)
+    }
+
+    /// get the kind of this metric
+    pub fn kind(&self) -> &MetricKind {
+        &self.kind
+    }
+
+    /// Add `n` to the metric's value
+    pub fn add(&self, n: usize) {
+        // relaxed ordering for operations on `value` poses no issues
+        // we're purely using atomic ops with no associated memory ops
+        self.value.fetch_add(n, Ordering::Relaxed);
+    }
+
+    /// Set the metric's value to `n`
+    pub fn set(&self, n: usize) {
+        self.value.store(n, Ordering::Relaxed);
+    }
+
+    /// What labels are present for this metric?
+    fn labels(&self) -> &[Label] {
+        &self.labels
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+/// How should the value of the metric be interpreted?
+pub enum MetricKind {
+    /// Number of output rows produced
+    OutputRows,
+    /// CPU time
+    CPUTime,
+    // TODO timestamp, etc
+    // https://github.com/apache/arrow-datafusion/issues/866
+    /// Arbitarary user defined type
+    Custom(Arc<str>),
```

Comment:
I wonder if this should just be `&'static str` and MetricKind made Copy

Reply:
done